Added support for page definitions

notKamui opened the pull request:
Closes #22 

Client can now submit a definition to received filtered and sorted pages. A definition is divided in two parts : 
- "filters" is an array (AND rules) of arrays (OR rules) of filter objects { field, operator, value [, type] }. (type is POS or NEG)
- "sorter" is an array of sort objects { field [, direction] } (direction is ASC or DES), ordered by precedence (meaning the first object has higher sort precedence over the second, etc)

notKamui committed:
Fixed file uploads

notKamui committed:
Use better type for request methods

notKamui committed:
Optimized async DB access

notKamui committed:
Properly checked exceptions in authentication validation

notKamui committed:
Added Filter and Sorter Definition DTOs

notKamui committed:
Added Filter and Sort DTOs

notKamui committed:
Added exceptions to manage unknown filters

notKamui committed:
Completed filter API

notKamui committed:
Linked filter API to the get file endpoint

notKamui committed:
Refactor

notKamui committed:
Restructured pageable API

notKamui committed:
Completed page sorting API

notKamui committed:
Users can now use page definitions

notKamui committed:
Refactor

notKamui committed:
Replaced ExceptionAdvisor API with the StatusPages plugin

notKamui committed:
Replaced ExceptionAdvisor API with the StatusPages plugin

notKamui committed:
Format

notKamui committed:
Format

notKamui committed:
Migrated client to Vite v3 for omega speedy build and hot reload

notKamui committed:
Updated project node version

notKamui committed:
Updated gradle dependencies

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/rest/core/pageable/DTO.kt:
```diff
@@ -10,4 +14,33 @@ class PageDTO<T : DTO>(
+    val direction: SortDirection = SortDirection.ASC,
+) : DTO
+
+enum class FilterType { POS, NEG }
```
Maybe use `POSITIVE` and `NEGATIVE` to be more explicit?

notKamui replied:
I wanted to keep the three letter scheme, but fair enough

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/rest/core/pageable/DTO.kt:
```diff
@@ -10,4 +14,33 @@ class PageDTO<T : DTO>(
+) : DTO
+
+enum class FilterType { POS, NEG }
+enum class SortDirection { ASC, DES }
```
Why not using `DESC` just like anyone else?

notKamui replied:
Same as POS and NEG

SlamaFR reviewed (changes requested):
Nice PR overall, just two minor points.

Also, I would like to see integration tests of this new feature, which is pretty critical in my opinion..

notKamui committed:
Changed filter and sort type flags to more regular ones

notKamui committed:
Updated gitignore

notKamui committed:
Added test data for files

notKamui committed:
Added test suite for files

notKamui committed:
Updated clean task

notKamui committed:
Trying to fix tests

notKamui commented:
This should be merged after #26

SlamaFR committed:
Update Jenkinsfile

notKamui committed:
Merge pull request #26 from Black-Kamelia/back/exceptions/ktorPlugin

Replaced ExceptionAdvisor API with the idiomatic StatusPages plugin

SlamaFR commented:
This PR looks good to me, although we'll need to add more tests for the filtering system.

notKamui committed:
Added mime types catalog

notKamui committed:
Added MimeTypes utility class

notKamui committed:
Added FileUtils utility class

notKamui committed:
Added AutoHead plugin

notKamui committed:
Moved QueryResult.kt to core

notKamui committed:
Changed user page route

notKamui committed:
Added ability to get and download a file

notKamui committed:
Added file tests

notKamui committed:
Added mime type to response header for file download (Not replacing Content-Type which should itself only be an octet-stream)

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/rest/file/Service.kt:
```diff
@@ -104,6 +96,14 @@ object FileService {
         return QueryResult.ok(Files.delete(fileId)?.toRepresentationDTO())
     }
+
+    suspend fun deleteFileByCodeAsAdmin(code: String): QueryResult<FileRepresentationDTO, List<ErrorDTO>> {
```
I think the `AsAdmin` part of the name, although relevant, is not needed here. We don't want to create extra long methods names that describe the method logic..

notKamui replied:
I do prefer clear names for what methods do in general ; there is virtually no cost over the benefit

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/rest/file/Service.kt:
```diff
@@ -1,79 +1,71 @@
-                ceil(total / pageSize.toDouble()).toLong(),
-                total
+    suspend fun getFiles(
+        user: User,
+        page: Long,
+        pageSize: Int,
+        definition: PageDefinitionDTO,
+    ): QueryResult<FilePageDTO, List<ErrorDTO>> {
+        val (files, total) = user.getFiles(page, pageSize, definition)
+        return QueryResult.ok(
+            FilePageDTO(
+                PageDTO(
+                    files.map { it.toRepresentationDTO() },
+                    page,
+                    pageSize,
+                    ceil(total / pageSize.toDouble()).toLong(),
+                    total
+                )
             )
-        ))
+        )
     }
```
I might be wrong, but with this method we never perform a permission check. I mean an admin could see private files. 

I think we should come up with a final convention of files visibility according to role. The owner should be able to see everything but an admin shouldn't. Let's keep a bit of privacy.

notKamui replied:
Okay fair enough

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/util/Extensions.kt:
```diff
@@ -6,18 +6,38 @@ import com.kamelia.jellyfish.core.ExpiredOrInvalidTokenException
+            .withParameter(ContentDisposition.Parameters.FileName, name)
+            .toString()
+    )
+    response.header("Mime-Type", type)
```
Shouldn't the header be `Content-Type` instead?

notKamui replied:
As said in the commit message, this is semantically a different thing. The Content-Type is always an octet-stream when sending a file that is supposed to be downloaded (it is to be interpreted as literal bytebuffer, without meaning), therefore, if we do still want to keep that information somewhere, we need another header.
Ktor call.sendFile automatically sets the Content-Type anyway.

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/util/FileUtils.kt:
```diff
@@ -0,0 +1,55 @@
+        return if (file.exists()) file else null
+    }
+
+    suspend fun write(
+        owner: UUID,
+        file: PartData.FileItem
+    ): Triple<String, String, Long> = withContext(Dispatchers.IO) {
```
This method absolutely needs documentation. Generally, we need to write documentation along with new methods so that anybody can understand quickly what's the point of every method.

SlamaFR commented on server/src/main/resources/mime-type-catalog.csv:
```diff
@@ -0,0 +1,989 @@
+.123,application/vnd.lotus-1-2-3
```
Note to myself: Add support of RAW pictures extensions (CR2, CRW, SR2, SRF, etc)

https://stackoverflow.com/questions/43473056/which-mime-type-should-be-used-for-a-raw-image

SlamaFR reviewed (changes requested):
I still have some points of interrogation.

notKamui committed:
Added missing mime type mappings

notKamui committed:
Restricted file access:
- pages: only the owner can view all their files, everyone else, even admins, can only see public files
- single: if the file is private, only the owner can download it, but anyone can download a public or unlisted file.

notKamui committed:
Added documentation

SlamaFR commented on server/src/main/kotlin/com/kamelia/jellyfish/rest/file/Routes.kt:
```diff
@@ -49,24 +59,45 @@ private fun Route.uploadFileFromToken() = postOrCatch(path = "/upload/token") {
+            val file = FileUtils.getOrNull(data!!.ownerId, code)
+            if (file != null) {
+                call.respondFile(file, data.name, data.mimeType)
+            } else {
+                FileService.deleteFileByCodeAsAdmin(code) // remove orphaned file
+                call.respond(QueryResult.notFound())
+            }
```
Not sure if this is a good idea. Such a rare event should not be handled automatically, instead we should come up with a report system, or simply count orphaned files discovered.

The goal is to tell admins there is a problem and let them handle it.

notKamui replied:
Noted, agreed, ill replace it with a todo comment ; we shall handle that later when we have admin routes

SlamaFR reviewed (changes requested):
Almost perfect! 😅

notKamui committed:
Hotfix
